internal/storage/inmem/watch.go

```diff
@@ -0,0 +1,188 @@
+package inmem
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/internal/storage"
+	"github.com/hashicorp/consul/proto-public/pbresource"
+	"github.com/hashicorp/consul/proto/private/pbsubscribe"
+)
+
+// Watch implements the storage.Watch interface using a stream.Subscription.
+type Watch struct {
+	sub   *stream.Subscription
+	query query
+
+	// events holds excess events when they are bundled in a stream.PayloadEvents,
+	// until Next is called again.
+	events []stream.Event
+}
+
+// Next returns the next WatchEvent, blocking until one is available.
+func (w *Watch) Next(ctx context.Context) (*pbresource.WatchEvent, error) {
+	for {
+		e, err := w.nextEvent(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		event := e.Payload.(eventPayload).event
+		if w.query.matches(event.Resource) {
+			return event, nil
+		}
+	}
+}
+
+func (w *Watch) nextEvent(ctx context.Context) (*stream.Event, error) {
+	if len(w.events) != 0 {
+		event := w.events[0]
+		w.events = w.events[1:]
+		return &event, nil
+	}
+
+	for {
+		e, err := w.sub.Next(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if e.IsFramingEvent() {
+			continue
+		}
+
+		switch t := e.Payload.(type) {
+		case eventPayload:
+			return &e, nil
+		case *stream.PayloadEvents:
+			if len(t.Items) == 0 {
+				continue
+			}
+
+			event, rest := t.Items[0], t.Items[1:]
+			w.events = rest
+			return &event, nil
+		}
+	}
+}
+
+var eventTopic = stream.StringTopic("resources")
+
+type eventPayload struct {
+	subject stream.Subject
+	event   *pbresource.WatchEvent
+}
+
+func (p eventPayload) Subject() stream.Subject { return p.subject }
+
+// These methods are required by the stream.Payload interface, but we don't use them.
+func (eventPayload) HasReadPermission(acl.Authorizer) bool         { return false }
+func (eventPayload) ToSubscriptionEvent(uint64) *pbsubscribe.Event { return nil }
+
+type wildcardSubject struct {
+	resourceType storage.UnversionedType
+}
+
+func (s wildcardSubject) String() string {
+	return s.resourceType.Group + indexSeparator +
+		s.resourceType.Kind + indexSeparator +
+		storage.Wildcard
+}
+
+type tenancySubject struct {
+	resourceType storage.UnversionedType
+	tenancy      *pbresource.Tenancy
+}
+
+func (s tenancySubject) String() string {
+	return s.resourceType.Group + indexSeparator +
+		s.resourceType.Kind + indexSeparator +
+		s.tenancy.Partition + indexSeparator +
+		s.tenancy.PeerName + indexSeparator +
+		s.tenancy.Namespace
+}
+
+// publishEvent sends the event to the relevant Watches.
+func (s *Store) publishEvent(idx uint64, event *pbresource.WatchEvent) {
```

Comment:
[super nit :) ] Does the caller need to construct a WatchEvent or could they do:
```
s.publishEvent(idx, pbresource.WatchEvent_OPERATION_UPSERT, res)
```

instead of 
```
	s.publishEvent(idx, &pbresource.WatchEvent{
		Operation: pbresource.WatchEvent_OPERATION_UPSERT,
		Resource:  res,
	})
```

Reply:
Good call! 🙌🏻

Reply:
Fixed in https://github.com/hashicorp/consul/pull/16538/commits/0167a5bc1eedd1d7ba6badae9a12ec366e7bb7dc